File: java-workshop/part-3-mocks-clean/src/test/java/service/ShippingServiceTest.java
package service;

import domain.ShippingConfirmation;
import gateway.ShippingGateway;
import helper.OrderBuilder;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class ShippingServiceTest {

    @Mock
    private ShippingGateway shippingGateway;

    @InjectMocks
    private GatewayShippingService shippingService;

    @Test
    void createShipment_returnsGatewayConfirmation() {
        var order = OrderBuilder.confirmed().withId("ORDER-001").build();
        var confirmation = new ShippingConfirmation("ORDER-001", "TRACK-123", "2025-12-25");
        when(shippingGateway.createShipment(any())).thenReturn(confirmation);

        var result = shippingService.createShipment(order);

        assertEquals(confirmation, result);
        verify(shippingGateway).createShipment(any());
    }

    @Test
    void cancelShipment_delegatesToGateway() {
        when(shippingGateway.cancelShipment(anyString())).thenReturn(true);

        assertTrue(shippingService.cancelShipment("TRACK-123"));
        verify(shippingGateway).cancelShipment("TRACK-123");
    }
}

File: csharp-workshop/Part3.MocksClean.Tests/Service/ShippingServiceTest.cs
using NSubstitute;
using Part3.MocksClean.Domain;
using Part3.MocksClean.Gateway;
using Part3.MocksClean.Service;
using Part3.MocksClean.Tests.Helper;
using Xunit;

namespace Part3.MocksClean.Tests.Service;

public class ShippingServiceTest
{
    private readonly IShippingGateway _shippingGateway = Substitute.For<IShippingGateway>();
    private readonly GatewayShippingService _shippingService;

    public ShippingServiceTest()
    {
        _shippingService = new GatewayShippingService(_shippingGateway);
    }

    [Fact]
    public void P3_CreateShipment_ReturnsGatewayConfirmation()
    {
        var order = OrderBuilder.Confirmed().WithId("ORDER-001").Build();
        var confirmation = new ShippingConfirmation("ORDER-001", "TRACK-123", "2025-12-25");
        _shippingGateway.CreateShipment(Arg.Any<ShippingRequest>()).Returns(confirmation);

        var result = _shippingService.CreateShipment(order);

        Assert.Equal(confirmation, result);
        _shippingGateway.Received(1).CreateShipment(Arg.Any<ShippingRequest>());
    }

    [Fact]
    public void P3_CancelShipment_DelegatesToGateway()
    {
        _shippingGateway.CancelShipment(Arg.Any<string>()).Returns(true);

        Assert.True(_shippingService.CancelShipment("TRACK-123"));
        _shippingGateway.Received(1).CancelShipment("TRACK-123");
    }
}
